### DISCUSSION TITLE
Model is dirty though array is the same

### DISCUSSION THREAD

**Discussion by jaulz** [spatie/laravel-schemaless-attributes#85] [Q&A]
I noticed that `updateOrCreate` triggers an update when passing a schemaless array though the arrays' content is the same. As far as I can see the array from Postgres has some spaces:
````php
["en-gb", "en-us", "en", "de"]
````

Whereas PHP's `json_encode` has no spaces:
````php
["en-gb","en-us","en","de"]
````

Does anyone have a smart idea to overcome this? A custom `json_encode` maybe?

**Comment by jaulz** [comment]
@freekmurze thanks for moving it but actually, I thought it is an issue or is it by design?

**Reply by dylanbr** [reply]
I am having the same issue.

My workaround is to override `originalIsEquivalent()` on the model, and for `SchemalessAttributes`, use `json_encode(json_decode($original))` to re-encode the original value, and then compare that with the current attribute value.

I am interested to hear if this discussion results in other ways to achieve this.

**Reply by danielbehrendt** [reply]
Same problem here using MySQL 8 which adds single spaces:

_MySQL also discards extra whitespace between keys, values, or elements in the original JSON document, and leaves (or inserts, when necessary) a single space following each comma (,) or colon (:) when displaying it. This is done to enhance readability._


This is really a problem when using schemaless attributes in combination with an auditing package as a new auditing entry gets created even if there where no changes made. 

Currently I do not have any workaround and I would appreciate any help!

**Reply by danielbehrendt** [reply]
@dylanbr would you mind sharing your workarounds (especially how an where you implemented them)?

**Reply by dylanbr** [reply]
As mentioned above, I'm overriding `originalIsEquivalent()`, a method on `Model`. This is called, with an attribute name as the parameter, when determining whether that attribute has changed.

I'm using a trait, although you could also add it to a base class. There is other unrelated code, so I am unable to paste it all here, but what it does is:

 - Check if the given attribute is cast to `SchemalessAttributes`
 - If so, take the original and current values, and run `json_decode()` followed by `json_encode()` on both, then compare the re-encoded values and return.
 - Otherwise return `parent::originalIsEquivalent()`.

Hope this helps.
